src/data/finvec.lean

```diff
@@ -0,0 +1,193 @@
+def finvec (n : ℕ) (α : Type*) : Type* := fin n → α
+
+namespace finvec
+
+variables {α : Type*}
+
+instance : unique (finvec 0 α) :=
+⟨⟨fin_zero_elim⟩, by { intro x, ext i, exact fin_zero_elim i }⟩
+
+/-- Transport a vector across an equality of dimensions.
+This is implemented without using `eq.rec` so that it will reduce
+when evaluated at a constructor `⟨j, h : j < n'⟩` of `fin n'`. -/
+protected def cast {n n' : ℕ} (h : n = n') : finvec n α → finvec n' α :=
+λ x j, x (fin.cast h.symm j)
+
+@[simp] lemma cast_app {n n' : ℕ} {h : n = n'} {x : finvec n α} {j : fin n'} :
+  (x.cast h) j = x (j.cast h.symm) :=
+rfl
+
+@[simp] lemma cast_id {n : ℕ} {h : n = n} {x : finvec n α} : x.cast h = x :=
+by { ext ⟨i, _⟩, refl }
+
+section prod
+
+/-! ### Products
+
+The isomorphism αⁿ⁺ᵐ ≅ αⁿ × αᵐ is ubiquitous in mathematics, and usually
+invoked implicitly. Lean maintains a distinction between the two sides of course.
+Here, we provide an API which presents `finvec (n+m) α` as though it were
+a structure with two fields `left : finvec n α` and `right : finvec m α`,
+with constructor `append : finvec n α → finvec m α → finvec (n+m) α`, and
+corresponding induction principle `prod_induction`.
+
+The arguments `n m : ℕ` are implicit throughout this API, to be inferred
+from the argument or result types; this may seem problematic, since a number
+can be written in the form `n+m` in many ways, but actually works well
+in practice. -/
+
+variables {n m : ℕ}
+
+/-- Project `αⁿ⁺ᵐ ≅ αⁿ × αᵐ` to the first factor. -/
+protected def left : finvec (n + m) α → finvec n α :=
+λ x j, x (fin.cast_add m j)
+
+/-- Project `αⁿ⁺ᵐ ≅ αⁿ × αᵐ` to the second factor. -/
+protected def right : finvec (n + m) α → finvec m α :=
+λ x j, x (fin.nat_add n j)
+
+/-- The canonical equivalence `αⁿ⁺ᵐ ≅ αⁿ × αᵐ`. -/
+def append_equiv : finvec (n + m) α ≃ finvec n α × finvec m α :=
+calc  (fin (n + m) → α)
+    ≃ (fin n ⊕ fin m → α)        : equiv.arrow_congr sum_fin_sum_equiv.symm (equiv.refl _)
+... ≃ (fin n → α) × (fin m → α)  : (equiv.sum_arrow_equiv_prod_arrow (fin n) (fin m) α)
+
+/-- The concatenation of two `finvec`s. -/
+def append (x : finvec n α) (y : finvec m α) : finvec (n + m) α :=
+append_equiv.symm (x, y)
+
+localized "infixr ` ++ ` := finvec.append" in finvec
+
+lemma append_equiv_app {z : finvec (n + m) α} : append_equiv z = (z.left, z.right) := rfl
+
+lemma append_equiv_symm_app {x : finvec n α} {y : finvec m α} :
+  append_equiv.symm (x, y) = x ++ y :=
+rfl
+
+@[simp] lemma left_append_right (x : finvec (n + m) α) : x.left ++ x.right = x :=
+append_equiv.symm_apply_apply x
+
+/-- Induction principle for `finvec (n + m) α`,
+imagined as a structure containing its left and right parts. -/
+@[elab_as_eliminator]
+protected lemma prod_induction {C : finvec (n + m) α → Prop}
+  (h : Π (x : finvec n α) (y : finvec m α), C (x ++ y)) (z : finvec (n + m) α) : C z :=
+left_append_right z ▸ h z.left z.right
```

Comment:
```suggestion
protected def prod_induction {C : finvec (n + m) α → Sort*}
  (h : Π (x : finvec n α) (y : finvec m α), C (x ++ y)) (z : finvec (n + m) α) : C z :=
by { rw ← left_append_right z, exact h z.left z.right }

lemma prod_induction_beta {C : finvec (n + m) α → Sort*}
  (h : Π (x : finvec n α) (y : finvec m α), C (x ++ y)) (x : finvec n α) (y : finvec m α) : 
  (x.append y).prod_induction h = h x y :=
by { apply eq_of_heq, refine (eq_mpr_heq _ _).trans _, rw [left_append, right_append] }
```
This must be below `left_append` and `right_append`. Potentially can be renamed to `prod_rec`.